lib/foreman_remote_execution_core/script_runner.rb

```diff
@@ -97,11 +97,32 @@ def close
 
     def session
       @session ||= begin
-                     @logger.debug("opening session to #{@ssh_user}@#{@host}")
-                     Net::SSH.start(@host, @ssh_user, ssh_options)
+                     with_process_reaping do
+                       @logger.debug("opening session to #{@ssh_user}@#{@host}")
+                       Net::SSH.start(@host, @ssh_user, ssh_options)
+                     end
                    end
     end
 
+    def with_process_reaping(&block)
+      yield
+    rescue => e
+      reap
+      raise e
+    end
+
+    def reap
+      reaped = 0
+      # While there are any dead child processes without waiting for the ones that are running
+      while Process.waitpid(-1, Process::WNOHANG)
```

Comment:
I'm a bit worried about this causing some race-conditions, when it would fall into a time where some process finished, but `io.close` was not called just yet. As a conseuqnece, the `io.close` would for example not set the `$?` variable when calling the proper `io.close`. I can't think of any other way than fixing it directly in net ssh